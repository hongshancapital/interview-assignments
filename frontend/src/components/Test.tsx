import React, {
  FC,
  useState,
  useCallback,
  ChangeEventHandler,
  ChangeEvent,
  useRef,
} from "react";

export const Test: FC = () => {
  const [gender, setGender] = useState("man");
  const onChange: ChangeEventHandler<HTMLInputElement> = useCallback((e) => {
    setGender(e.target.value);
  }, []);

  const [name, setName] = useState<string>("");
  const onInputChange: ChangeEventHandler<HTMLInputElement> = useCallback(
    (e) => {
      console.log(e.target.value);
      setName(e.target.value);
    },
    []
  );

  const radioRef = useRef<HTMLDivElement>(null);
  const nameRef = useRef<HTMLDivElement>(null);

  const verifyAll = useCallback(() => {
    let res = true;
    let first: HTMLDivElement | null = null;

    const setFirst = (v: HTMLDivElement | null) => {
      if (!first) {
        first = v;
      }
      return first;
    };

    if (!gender) {
      res = false;
      first = setFirst(radioRef.current);
    }
    if (!name) {
      res = false;

      first = setFirst(nameRef.current);
    }

    first?.scrollIntoView();

    return res;
  }, [gender, name]);

  const handleSubmit = useCallback(() => {
    if (!verifyAll()) return;
    console.log("提交");
  }, [verifyAll]);

  return (
    <div>
      <div>
        <div ref={radioRef}>
          <input
            id="man"
            type="radio"
            name="gender"
            value="man"
            checked={gender === "man"}
            onChange={onChange}
          />
          <label htmlFor="man">男</label>
        </div>

        <div>
          <input
            id="female"
            type="radio"
            name="gender"
            checked={gender === "female"}
            value={"female"}
            onChange={onChange}
          />
          <label htmlFor="female">女</label>
        </div>
      </div>

      <div style={{ height: 800 }}></div>

      <div ref={nameRef}>
        <input value={name} onChange={onInputChange}></input>
      </div>

      <div style={{ height: 800 }}></div>

      <button onClick={handleSubmit}>提交</button>
    </div>
  );
};

