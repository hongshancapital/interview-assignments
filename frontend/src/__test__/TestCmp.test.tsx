import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Test from '../pages/Test';

describe('Test Component test', () => {
  it('Test Component rendered', () => {
    const { container } = render(<Test />);
    expect(container).toMatchSnapshot();
  });

  it('Test Component contains three carouselItem', () => {
    const { getByTestId } = render(<Test />);
    const carouselInner = getByTestId('carousel-inner-wrapper');
    const indicator = getByTestId('indicator-wrapper');
    expect(carouselInner).toBeInTheDocument();
    expect(carouselInner.childElementCount).toEqual(3);
    expect(indicator).toBeInTheDocument();
    expect(indicator.childElementCount).toEqual(3);
  });

  it('click last Indicator', () => {
    const { getByTestId } = render(<Test />);
    const carouselInner = getByTestId('carousel-inner-wrapper');
    const indicator = getByTestId('indicator-wrapper');
    fireEvent.click(indicator.lastChild!);
    expect(carouselInner).toHaveStyle({ transform: 'translateX(-200%)' });
  });
});
