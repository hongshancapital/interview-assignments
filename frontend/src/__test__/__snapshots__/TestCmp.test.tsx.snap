// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Component test Test Component rendered 1`] = `
<div>
  <div
    class="wrapper"
  >
    <div
      class="innerWrapper"
      data-testid="carousel-inner-wrapper"
      style="transform: translateX(-0%);"
    >
      <div
        class="carouselItem"
      >
        <div
          class="itemWrapper"
        >
          <div
            class="textWrapper"
            style="color: white;"
          >
            <div
              class="title"
            >
              xPhone
            </div>
            <div
              class="desc"
            >
              Lots to love. Less to spend.
            </div>
            <div
              class="desc"
            >
              Starting at $399
            </div>
          </div>
          <div
            class="imgWrapper"
          >
            <img
              alt=""
              src="iphone.png"
            />
          </div>
        </div>
      </div>
      <div
        class="carouselItem"
      >
        <div
          class="itemWrapper"
        >
          <div
            class="textWrapper"
          >
            <div
              class="title"
            >
              Tablet
            </div>
            <div
              class="desc"
            >
              Just the right amount of everything
            </div>
          </div>
          <div
            class="imgWrapper"
          >
            <img
              alt=""
              src="tablet.png"
            />
          </div>
        </div>
      </div>
      <div
        class="carouselItem"
      >
        <div
          class="itemWrapper"
        >
          <div
            class="textWrapper"
          >
            <div
              class="title"
            >
              Buy a Tablet or xPhone for college. Get airPods
            </div>
          </div>
          <div
            class="imgWrapper"
          >
            <img
              alt=""
              src="airpods.png"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="wrapper"
      data-testid="indicator-wrapper"
    >
      <div
        class="item"
      >
        <div
          class="activeItem"
          style="transition-duration: 3000ms;"
        />
      </div>
      <div
        class="item"
      >
        <div
          class=""
        />
      </div>
      <div
        class="item"
      >
        <div
          class=""
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Test Component test click Indicator of Carousel  1`] = `
<div>
  <div
    class="wrapper"
  >
    <div
      class="innerWrapper"
      data-testid="carousel-inner-wrapper"
      style="transform: translateX(-200%);"
    >
      <div
        class="carouselItem"
      >
        <div
          class="itemWrapper"
        >
          <div
            class="textWrapper"
            style="color: white;"
          >
            <div
              class="title"
            >
              xPhone
            </div>
            <div
              class="desc"
            >
              Lots to love. Less to spend.
            </div>
            <div
              class="desc"
            >
              Starting at $399
            </div>
          </div>
          <div
            class="imgWrapper"
          >
            <img
              alt=""
              src="iphone.png"
            />
          </div>
        </div>
      </div>
      <div
        class="carouselItem"
      >
        <div
          class="itemWrapper"
        >
          <div
            class="textWrapper"
          >
            <div
              class="title"
            >
              Tablet
            </div>
            <div
              class="desc"
            >
              Just the right amount of everything
            </div>
          </div>
          <div
            class="imgWrapper"
          >
            <img
              alt=""
              src="tablet.png"
            />
          </div>
        </div>
      </div>
      <div
        class="carouselItem"
      >
        <div
          class="itemWrapper"
        >
          <div
            class="textWrapper"
          >
            <div
              class="title"
            >
              Buy a Tablet or xPhone for college. Get airPods
            </div>
          </div>
          <div
            class="imgWrapper"
          >
            <img
              alt=""
              src="airpods.png"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="wrapper"
      data-testid="indicator-wrapper"
    >
      <div
        class="item"
      >
        <div
          class=""
          style=""
        />
      </div>
      <div
        class="item"
      >
        <div
          class=""
        />
      </div>
      <div
        class="item"
      >
        <div
          class="activeItem"
          style="transition-duration: 3000ms;"
        />
      </div>
    </div>
  </div>
</div>
`;
